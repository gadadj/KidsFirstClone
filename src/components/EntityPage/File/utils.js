import _ from 'lodash';
import { differenceInYears, differenceInDays, addYears } from 'date-fns';

export const pickData = (data, valuePath, transform) => {
  const selectedData = _.get(data, valuePath, null);
  if (transform) {
    return transform(selectedData);
  }
  return selectedData === null ? '--' : selectedData;
};

export const formatDate = date => {
  const now = new Date();
  const yearDiff = differenceInYears(now, date);

  // get diff in days, only from months and days, not years
  const dateSameYear = addYears(date, yearDiff);
  const dayDiff = differenceInDays(now, dateSameYear);
  return `${yearDiff} years ${dayDiff} days`;
};

export const formatToGB = size => `${(size / 1000000000).toFixed(2)} GB`;
