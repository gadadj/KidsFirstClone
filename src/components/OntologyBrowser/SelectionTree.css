.ant-tree.hide-file-icon .ant-tree-switcher-line-icon span.ant-tree-switcher-noop {
  background: transparent;
}

.ant-tree.ant-tree-show-line.hide-file-icon span.ant-tree-switcher-noop {
  color: transparent;
  height: 0px;
}

.ant-tree .ant-tree-title span.ant-tag {
  color: rgba(0, 0, 0, 0.65);
}

.ant-tree .ant-tree-list .ant-tree-treenode {
  width: 100%;
}

.ant-tree .ant-tree-treenode .ant-tree-node-content-wrapper {
  flex-grow: 1;
}

.ant-tree .ant-tree-treenode .ant-tree-title {
  display: flex;
  justify-content: space-between;
}

.ant-tree .ant-tree-list {
  overflow-y: hidden;
}

.ant-tree .ant-tree-title span.ant-tag.label-document-count {
  background-color: #e0e0e0;
  height: fit-content;
}

.ant-tree-node-content-wrapper .highlight {
  background-color: #f7ed9c;
}
