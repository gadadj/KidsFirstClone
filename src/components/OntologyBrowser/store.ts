import { graphql } from '../../services/arranger';
import { TreeNode } from 'antd/lib/tree-select';

type PhenotypeSource = {
  key: string;
  doc_count: number;
  top_hits: {
    parents: string[];
  };
  matches?: number;
};

export type TreeNode = {
  title: React.ReactElement | string;
  text: string;
  key: string;
  hasChildren?: boolean;
  children: TreeNode[];
  results?: number;
  hidden?: boolean;
  depth?: number;
};

export class PhenotypeStore {
  // Flat representation of phenotype from graphql source
  phenotypes: PhenotypeSource[] = [];
  // Tree of Phenotype Node
  tree: TreeNode[] = [];

  fetch = (sqon?: any) => {
    this.phenotypes = [];
    this.tree = [];
    return this.getPhenotypes(sqon).then((data: PhenotypeSource[]) => {
      const stripedData = this.remoteSingleRootNode(data);
      this.phenotypes = stripedData;
      this.tree = this.generateTree();
      return true;
    });
  };

  createNodeFromSource = (
    ontologySource: PhenotypeSource,
    parent?: TreeNode,
    depth: number = 0,
  ) => ({
    title: ontologySource.key,
    text: ontologySource.key,
    key: parent ? `${parent.key}-${ontologySource.key}` : ontologySource.key,
    children: [],
    results: ontologySource.doc_count,
    value: ontologySource.doc_count,
    name: ontologySource.key,
    depth,
  });

  private populateNodeChild = (treeNode: TreeNode, source: PhenotypeSource, depth: number = 0) => {
    this.phenotypes.forEach((phenotypeSource: PhenotypeSource) => {
      if (phenotypeSource.top_hits.parents.includes(source.key)) {
        let childNode = this.createNodeFromSource(phenotypeSource, treeNode, depth);
        treeNode.children.push(this.populateNodeChild(childNode, phenotypeSource, depth + 1));
      }
    });
    return treeNode;
  };

  generateTree = () => {
    let workingTree: TreeNode[] = [];
    let workingPhenotypes = [...this.phenotypes];

    workingPhenotypes.forEach((sourcePhenotype) => {
      let phenotype: TreeNode;
      // start from root and then look for each element inhereting from that node
      if (!sourcePhenotype.top_hits.parents.length) {
        phenotype = this.createNodeFromSource(sourcePhenotype);
        workingTree.push(this.populateNodeChild(phenotype, sourcePhenotype, 1));
      }
    });
    return workingTree;
  };

  getTreeNodeForKey = (key: string, treeNode = this.tree): TreeNode | null => {
    let result: TreeNode | null = null;
    for (let i = 0; i < treeNode.length; i++) {
      if (treeNode[i].key === key) {
        result = treeNode[i];
        break;
      }
      result = this.getTreeNodeForKey(key, treeNode[i].children);
      if (result) {
        break;
      }
    }
    return result;
  };

  getChildrenKeys = (node: TreeNode, root = false) => {
    let nKeys: string[] = [];
    node.children.forEach((i) => {
      nKeys = nKeys.concat(this.getChildrenKeys(i));
    });
    if (!root) {
      nKeys.push(node.key);
    }
    return nKeys;
  };

  getTree = (maxDepth: number = 2) => {
    if (this.tree.length === 0) return [];

    const newTree = [...this.tree];
    const cleanTree = (node: TreeNode) => {
      if (!node.depth) return;
      if (node.depth >= maxDepth) {
        delete node.children;
      } else if (node.children.length > 0) {
        for (let n of node.children) {
          cleanTree(n);
        }
      }
    };
    cleanTree(newTree[0]);
    return newTree;
  };

  buildPhenotypeQuery = () => `query($sqon: JSON) {
    participant {
      aggregations(filters: $sqon, aggregations_filter_themselves: false) {
        observed_phenotype__name {
          buckets{
            key,
            doc_count,
            top_hits(_source: "observed_phenotype.parents", size: 1)
          }
        }
      }
    }
  }
  `;

  getPhenotypes = async (sqon?: any) => {
    const body = {
      query: this.buildPhenotypeQuery(),
      variables: JSON.stringify({
        sqon: sqon,
      }),
    };
    try {
      const { data } = await graphql()(body);
      return data.data.participant.aggregations.observed_phenotype__name.buckets;
    } catch (error) {
      console.warn(error);
      return [];
    }
  };

  remoteSingleRootNode = (phenotypes: PhenotypeSource[]) => {
    return phenotypes
      .map((p) => (p.key !== 'All (HP:0000001)' ? p : null))
      .filter((p): p is PhenotypeSource => p !== null)
      .map((p) => {
        const index = p.top_hits.parents.indexOf('All (HP:0000001)');
        if (!index) {
          p.top_hits.parents.splice(index, 1);
        }
        return p;
      });
  };
}
