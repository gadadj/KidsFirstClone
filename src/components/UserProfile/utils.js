import get from 'lodash/get';
import { ROLES } from 'common/constants';
import React, { Fragment } from 'react';
import { SocialIcon } from 'react-social-icons';
import { kfFacebook, kfGithub, kfTwitter } from 'common/injectGlobals';
import orchidIcon from 'assets/icon-findemeon-orchid.png';
import WebsiteIcon from 'icons/WebsiteIcon';
import GoogleScholarIcon from 'icons/GoogleScholarIcon';
import LinkedInIcon from 'icons/LinkedInIcon';
import { findMeFields } from './constants';
import style from 'components/UserProfile/style';
import { CheckOutlined, EditOutlined } from '@ant-design/icons';
import { Button, Typography } from 'antd';

const { Text } = Typography;

const addWebProtocolToUrlIfNeeded = value => {
  if (!value || value.startsWith('http://') || value.startsWith('https://')) {
    return value;
  }
  return `https://${value}`;
};

const href = v => addWebProtocolToUrlIfNeeded(v);

export const socialItems = (width = 32, height = 32) => {
  const commonSize = { height, width };
  return {
    website: {
      icon: <WebsiteIcon {...commonSize} />,
      name: 'Website URL',
      placeholder: 'e.g. kidsfirstdrc.org',
      type: 'text',
      service: 'Website',
      href,
    },
    googleScholarId: {
      icon: <GoogleScholarIcon {...commonSize} />,
      name: 'Google Scholar URL',
      placeholder: 'e.g. scholar.google.com/citations?user=CsD2_4MAAAAJ',
      type: 'text',
      service: 'Google Scholar',
      href,
    },
    linkedin: {
      icon: <LinkedInIcon {...commonSize} />,
      name: 'LinkedIn URL',
      placeholder: 'e.g. linkedin.com/in/acresnick',
      type: 'text',
      service: 'LinkedIn',
      href,
    },
    facebook: {
      icon: <SocialIcon url={kfFacebook} style={{ ...commonSize }} />,
      name: 'Facebook URL',
      placeholder: 'e.g. facebook.com/kidsfirstDRC',
      type: 'text',
      service: 'Facebook',
      href,
    },
    twitter: {
      icon: <SocialIcon url={kfTwitter} style={{ ...commonSize }} />,
      name: 'Twitter handle/username',
      placeholder: 'e.g. @kidsfirstDRC',
      type: 'text',
      href: v => `https://twitter.com/${v}`,
      linkText: v => `@${v}`,
      service: 'Twitter',
    },
    github: {
      icon: <SocialIcon url={kfGithub} style={{ ...commonSize }} />,
      name: 'Github username',
      placeholder: 'e.g. kids-first',
      type: 'text',
      href: v => `https://github.com/${v}`,
      service: 'Github',
    },
    orchid: {
      icon: <img alt="ORCHID" src={orchidIcon} {...commonSize} />,
      name: 'ORCID ID',
      placeholder: 'e.g. 0000-0003-0436-4189',
      type: 'text',
      href: v => `https://orcid.org/${v}`,
      service: 'Orchid',
    },
  };
};
export const userProfileBackground = (
  profile,
  { showBanner = true, gradientDirection = 'right' } = {},
) => {
  const role = ROLES.find(x => x.type === get(profile, 'roles[0]', '')) || {};
  const banner = get(role, 'banner', '');
  const profileColors = get(role, 'profileColors', {});
  return {
    position: 'relative',
    backgroundPositionX: 'right',
    backgroundRepeat: 'no-repeat',
    backgroundImage: ` ${showBanner ? `url(${banner}), ` : ``}
      linear-gradient(
        to ${gradientDirection},
        ${profileColors.gradientDark} 33%,
        ${profileColors.gradientMid} 66%,
        ${profileColors.gradientLight}
      )`,
  };
};

export const extractFindMeFromProfile = (profile = {}) => Object.entries(profile).reduce((accFindMe, [profileKey, profileValue]) => {
    if (findMeFields.includes(profileKey) && Boolean(profileValue)) {
      return { ...accFindMe, [profileKey]: profileValue };
    }
    return accFindMe;
  }, {});

export const isResearcher = data => data.roles[0] === 'research';

export const isCommunity = data => data.roles[0] === 'community';

export const showInstitution = data => isResearcher(data) || isCommunity(data);

export const makeCommonCardPropsReadOnly = ({
  isProfileUpdating,
  title,
  onClickEditCb,
  canEdit,
}) => ({
    loading: isProfileUpdating,
    title: <Text className={'header-title'}>{title}</Text>,
    className: 'card-container',
    headStyle: style.cardHeadStyle,
    bodyStyle: style.cardBodyStyle,
    extra: canEdit ? (
      <Button
        size={'small'}
        icon={<EditOutlined />}
        shape="round"
        onClick={onClickEditCb}
        style={{ backgroundColor: 'rgb(144, 38, 142)', color: 'white' }}
      >
        EDIT
      </Button>
    ) : null,
  });

export const makeCommonCardPropsEditing = ({
  isProfileUpdating,
  title,
  onClickCancelCb,
  disableSaveButton,
}) => ({
    loading: isProfileUpdating,
    title: <Text className={'header-title'}>{title}</Text>,
    className: 'card-container',
    headStyle: style.cardHeadStyle,
    bodyStyle: style.cardBodyStyleWhenEditing,
    extra: (
      <Fragment>
        <Button
          size={'small'}
          className={'extra-button'}
          shape="round"
          onClick={onClickCancelCb}
          style={{ color: 'rgb(144, 38, 142)' }}
        >
          CANCEL
        </Button>
        <Button
          size={'small'}
          className={'extra-button'}
          icon={<CheckOutlined />}
          shape="round"
          style={{
            backgroundColor: disableSaveButton ? 'lightgrey' : 'rgb(144, 38, 142)',
            color: 'white',
          }}
          disabled={Boolean(disableSaveButton)}
          htmlType="submit"
        >
          SAVE
        </Button>
      </Fragment>
    ),
  });

export const showWhenHasDataOrCanEdit = (data, canEdit) => Boolean(data) || canEdit;

export const hasFieldInError = fields =>
  Object.entries(fields || {}).some(([, value]) => Array.isArray(value) && value.length > 0);

//When persona throws an AccessError, show details to user.
export const getMsgFromAccessError = (rawError, defaultMsgIfNotFound = '') =>
  rawError?.response?.data?.message || defaultMsgIfNotFound;
