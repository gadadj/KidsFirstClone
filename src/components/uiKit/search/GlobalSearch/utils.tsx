import { isEmpty } from 'lodash';

export const highlightSearchMatch = (value: string, matchRegex: RegExp, search: string) => {
  const result = !isEmpty(search) && !isEmpty(value) && value.search(matchRegex) >= 0;

  return result ? (
    <span
      dangerouslySetInnerHTML={{
        __html: value.replace(matchRegex, (match) => `<strong>${match}</strong>`),
      }}
    ></span>
  ) : (
    value
  );
};
