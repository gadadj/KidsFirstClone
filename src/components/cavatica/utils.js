import { graphql } from 'services/arranger';
import { toGqlString } from 'services/utils';

export const getStudiesAggregations = ({ api, studies, sqon }) => {
  return !studies.length
    ? []
    : graphql(api)({
        query: `
          query approvedStudyAggs(${studies
            .map(study => `$${toGqlString(study)}_sqon: JSON`)
            .join(' ')}) {
            file {
              ${studies
                .map(
                  study => `
                  ${toGqlString(study)}: aggregations (filters: $${toGqlString(study)}_sqon){
                    latest_did {
                      buckets {
                        key
                      }
                    }
                    participants__study__name {
                      buckets {
                        key
                      }
                    }
                  }
                `,
                )
                .join(' ')}
            }
          }
        `,
        variables: studies.reduce(
          (acc, study) => ({
            ...acc,
            [`${toGqlString(study)}_sqon`]: {
              ...sqon,
              content: [
                ...sqon.content,
                {
                  op: 'in',
                  content: {
                    field: 'participants.study.external_id',
                    value: study,
                  },
                },
              ],
            },
          }),
          {},
        ),
      }).then(({ data: { file: fileAggs } }) =>
        Object.entries(fileAggs).map(([study, aggs]) => ({
          study: study,
          files: aggs.latest_did.buckets.map(({ key }) => key),
          studyName: aggs.participants__study__name.buckets.length
            ? aggs.participants__study__name.buckets[0].key
            : null,
        })),
      );
};
export const getUnapprovedStudiesForFiles = ({ api, files, approvedStudyAggs }) =>
  graphql(api)({
    query: `
    query ($sqon: JSON) {
      file {
        aggregations (filters: $sqon){
          participants__study__external_id {
            buckets {
              key
            }
          }
        }
      }
    }
  `,
    variables: {
      sqon: {
        op: 'and',
        content: [
          {
            op: 'in',
            content: {
              field: '_id',
              value: files,
            },
          },
        ],
      },
    },
  }).then(
    ({ data: { file: { aggregations: { participants__study__external_id: { buckets } } } } }) =>
      buckets
        .map(({ key }) => key)
        .filter(study => !approvedStudyAggs.map(({ study }) => study).includes(study)),
  );
