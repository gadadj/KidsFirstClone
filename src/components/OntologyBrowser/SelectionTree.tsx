import * as React from 'react';
import { Component, Fragment } from 'react';
import { Button, Col, Input, Row, Tag, Tree } from 'antd';
import { TreeNode } from './store';

import './SelectionTree.css';

type SelectionTreeProps = {
  dataSource: TreeNode[];
  checkedKeys: Array<string>;
  onItemSelect: Function;
  targetKeys: Array<string>;
  onItemSelectAll: Function;
};

type SelectionTreeState = {
  treeData: TreeNode[];
  expandedKeys: string[];
};

const AUTO_EXPAND_TREE = 2;
const MIN_SEARCH_TEXT_LENGTH = 3;

const getInitialKeysForExpand = (data: TreeNode[], collectedKeys: string[] = [], counter = 1) => {
  if (counter < AUTO_EXPAND_TREE) {
    data.forEach((node) => {
      counter++;
      collectedKeys.push(node.key);
      if (node.children) {
        getInitialKeysForExpand(node.children, collectedKeys, counter);
      }
    });
  }
  return collectedKeys;
};

export class SelectionTree extends Component<SelectionTreeProps, SelectionTreeState> {
  state = {
    treeData: [],
    expandedKeys: [],
  };

  componentDidMount() {
    const { dataSource } = this.props;
    this.setState({
      treeData: dataSource,
      expandedKeys: getInitialKeysForExpand(dataSource),
    });
  }

  generateTree = (
    treeNodes: TreeNode[] = [],
    checkedKeys: string[] = [],
    targetKeys: string[] = [],
    disabled: boolean = false,
  ): TreeNode[] => {
    return treeNodes
      .map(({ children, key, title, results, hidden }: TreeNode) => {
        const renderedTitle = (
          <Fragment>
            <span>{title}</span>
            <Tag className="label-document-count">{results}</Tag>
          </Fragment>
        );
        const isDisabled = targetKeys.includes(key || '') || disabled;
        const childrenShouldBeDisabled = checkedKeys.includes(key) || isDisabled;
        return {
          key: key,
          title: renderedTitle,
          text: title,
          disabled: isDisabled,
          children: this.generateTree(children, checkedKeys, targetKeys, childrenShouldBeDisabled),
          hasChildren: true,
          results: 324,
          hidden
        } as TreeNode;
      })
      .filter((node) => (node.hidden ? false : !node.hidden));
  };

  isChecked = (selectedKeys: Array<string>, eventKey: string | number) =>
    selectedKeys.indexOf(eventKey.toString()) !== -1;

  onChange = (e: React.ChangeEvent<HTMLInputElement>, treeData: TreeNode[]) => {
    const hits: string[] = [];
    let newExpandNode: string[] = [];

    if (e.target.value.length >= MIN_SEARCH_TEXT_LENGTH) {
      treeData.forEach((node) => this.searchInTree(e.target.value, node, hits));
      newExpandNode = hits;
    } else {
      treeData.forEach((node) => this.showAll(node));
      newExpandNode = getInitialKeysForExpand(treeData);
    }

    this.setState({
      treeData: treeData,
      expandedKeys: newExpandNode,
    });
  };

  searchInTree = (searchText: string, treeNode: TreeNode, hitTreeNodes: string[] = []) => {
    const cleanSearchText = searchText.replace(/[-/\\^$*+?.()|[\]{}]/g, '');
    const regex = new RegExp('\\b(\\w*' + cleanSearchText + '\\w*)\\b', 'gi');
    const text = treeNode.text;
    const result = text.search(regex) >= 0;
    let match = cleanSearchText === '' || result;

    if (treeNode.children.length > 0) {
      let matchChild = cleanSearchText === '' || false;
      treeNode.children.forEach((child: TreeNode) => {
        if (this.searchInTree(cleanSearchText, child, hitTreeNodes)) {
          matchChild = true;
        }
      });
      match = matchChild || match;
    }
    treeNode.hidden = !match;

    if (!treeNode.hidden) {
      hitTreeNodes.push(treeNode.key);
      if (result) {
        const [before, hit, after] = treeNode.text.split(regex);
        treeNode.title = (
          <span>
            {before}
            <div className={'highlight'} style={{ display: 'inherit' }}>
              {hit}
            </div>
            {after}
          </span>
        );
      }
    }
    return match;
  };

  showAll = (treeNode: TreeNode) => {
    treeNode.hidden = false;
    treeNode.title = treeNode.text;

    if (treeNode.children.length > 0) {
      treeNode.children.forEach((child: TreeNode) => {
        this.showAll(child);
      });
    }
  };

  onExpand = (expand: (string | number)[], info: Object) => {
    this.setState({
      expandedKeys: expand.map((v) => v.toString()),
    });
  };

  render() {
    const { checkedKeys, dataSource, onItemSelect, targetKeys, onItemSelectAll } = this.props;
    const { expandedKeys } = this.state;
    return (
      <Fragment>
        <Col style={{ position: 'sticky', top: 0, zIndex: 2, backgroundColor: '#fff' }}>
          <Row>
            <Input
              placeholder="Search for ontology term - Min 3 characters"
              onChange={(e) => this.onChange(e, dataSource)}
              allowClear
            />
          </Row>
          <Row justify="start">
            <Button
              type="link"
              onClick={(e) => {
                e.preventDefault();
                onItemSelectAll(checkedKeys, false);
              }}
            >
              Clear
            </Button>
          </Row>
        </Col>
        <Tree
          className="hide-file-icon"
          treeData={this.generateTree(dataSource, checkedKeys, targetKeys)}
          defaultExpandAll
          selectedKeys={[]}
          showLine
          showIcon={false}
          checkable
          onCheck={(_, { node }) => {
            const isChecked = !this.isChecked(checkedKeys, node.key);
            onItemSelect(node.key, isChecked);
          }}
          checkedKeys={checkedKeys}
          onSelect={(_, { node: { key } }) => {
            onItemSelect(key, !this.isChecked(checkedKeys, key));
          }}
          expandedKeys={expandedKeys}
          onExpand={this.onExpand}
        />
      </Fragment>
    );
  }
}
