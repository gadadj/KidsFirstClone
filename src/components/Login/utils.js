import get from 'lodash/get';
import isArrayLikeObject from 'lodash/isArrayLikeObject';
import toLower from 'lodash/toLower';
import jwtDecode from 'jwt-decode';

import { getProfile, createProfile } from 'services/profiles';
import { getUser as getCavaticaUser } from 'services/cavatica';
import { getAccessToken } from 'services/fence';
import { createExampleQueries } from 'services/riffQueries';
import { FENCES, CAVATICA } from 'common/constants';
import { store } from 'store';
import { loginSuccess, loginFailure } from 'store/actionCreators/user';

const initProfile = async (api, user, egoId) => {
  const profileCreation = createProfile(api)({ ...user, egoId });
  const sampleQueryCreation = createExampleQueries(api, egoId);
  const [x] = await Promise.all([profileCreation, sampleQueryCreation]);
  return x;
};

export const isAdminToken = ({ validatedPayload }) => {
  if (!validatedPayload) return false;
  const jwtUser = get(validatedPayload, 'context.user', {});
  const roles = jwtUser.roles;
  const type = jwtUser.type;
  // maintain backward compatibility
  return roles && null !== roles && isArrayLikeObject(roles)
    ? roles.includes('ADMIN')
    : type && type !== null && type === 'ADMIN';
};

export const getUserGroups = ({ validatedPayload }) => {
  if (!validatedPayload) return [];
  const jwtUser = get(validatedPayload, 'context.user', {});
  const groups = jwtUser.groups;
  return groups ? groups : [];
};

export const validateJWT = ({ jwt }) => {
  if (!jwt) return false;
  const validatedPayload = jwtDecode(jwt);
  const isCurrent = new Date(validatedPayload.exp * 1000).valueOf() > Date.now();
  const status = get(validatedPayload, 'context.user.status', '');
  const isApproved =
    isAdminToken({ validatedPayload }) || [status].map(toLower).includes('approved');
  return isCurrent && isApproved && validatedPayload;
};

export const handleJWT = async ({ provider, jwt, onFinish, setToken, setUser, api }) => {
  const jwtData = validateJWT({ jwt });

  if (!jwtData) {
    setToken();
    // clear the user details in the redux store
    store.dispatch(loginFailure());
  } else {
    try {
      await setToken({ token: jwt, provider });
      const user = jwtData.context.user;
      const egoId = jwtData.sub;
      const existingProfile = await getProfile(api)();
      const newProfile = !existingProfile ? await initProfile(api, user, egoId) : {};
      const loggedInUser = {
        ...(existingProfile || newProfile),
        email: user.email,
        egoGroups: user.groups,
      };
      await setUser({ ...loggedInUser, api });
      onFinish && onFinish(loggedInUser);

      // store the user details in the redux store
      store.dispatch(loginSuccess(loggedInUser));
    } catch (err) {
      console.error('Error during login', err);
      // clear the user details in the redux store
      store.dispatch(loginFailure());
    }
  }
  return jwtData;
};

/**
 * fetchIntegrationTokens
 * For all SERVICES listed in common/constants, call the key-store to retrieve any keys stored
 *  for the user.
 * Each call to key-store is resolved separately and asynchronously. Their value will be added
 *  to state once returned.
 */
export const fetchIntegrationTokens = ({ setIntegrationToken, api }) => {
  getCavaticaUser()
    .then((userData) => {
      setIntegrationToken(CAVATICA, JSON.stringify(userData));
    })
    .catch((error) => {
      console.error(error);
      // Could not retrieve cavatica user info, nothing to do.
    });

  // Get Gen3 Secret here
  FENCES.forEach((fence) => {
    getAccessToken(api, fence)
      .then((key) => {
        setIntegrationToken(fence, key);
      })
      .catch((res) => {
        console.error('Error getting Gen3 API Key');
        console.error(res);
      });
  });
};
