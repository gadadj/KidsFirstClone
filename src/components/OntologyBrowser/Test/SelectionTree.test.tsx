import { TreeNode } from '../store';
import { SelectionTree } from '../SelectionTree';
import * as React from 'react';
import Enzyme, { mount, shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { treeData } from './mockData';
import { jestPatchMatchMedia } from '../../../utils';

Enzyme.configure({ adapter: new Adapter() });

describe('In SelectionTree', () => {
  const checkedKeys: string[] = [];
  const targetKeys: string[] = [];
  const onItemSelect = jest.fn();
  const onItemSelectAll = jest.fn();
  const dataSource: TreeNode[] = treeData;

  beforeAll(() => jestPatchMatchMedia());

  let wrapper: ShallowWrapper<
    {},
    {
      treeData: TreeNode[];
      expandedKeys: string[];
    },
    SelectionTree
  >;

  const wrapperInstance = shallow<SelectionTree, {}, {}>(
    <SelectionTree
      checkedKeys={[]}
      dataSource={[]}
      onItemSelect={jest.fn()}
      onItemSelectAll={jest.fn()}
      targetKeys={[]}
      selectedField={''}
    />,
  );

  beforeEach(() => {
    wrapper = shallow<
      SelectionTree,
      {},
      {
        treeData: TreeNode[];
        expandedKeys: string[];
      }
    >(
      <SelectionTree
        dataSource={dataSource}
        checkedKeys={checkedKeys}
        onItemSelect={onItemSelect}
        targetKeys={targetKeys}
        onItemSelectAll={onItemSelectAll}
        selectedField={''}
      />,
    );
  });

  it('should renders selection tree', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should return expected result', () => {
    const expectedResult: TreeNode[] = [
      {
        title: 'Abnormality of the integument (HP:0001574)',
        text: 'Abnormality of the integument (HP:0001574)',
        key: 'Abnormality of the integument (HP:0001574)',
        hidden: false,
        results: 12,
        exactTagCount: 0,
        children: [
          {
            title: 'Abnormality of the skin (HP:0000951)',
            text: 'Abnormality of the skin (HP:0000951)',
            key: 'Abnormality of the integument (HP:0001574)-Abnormality of the skin (HP:0000951)',
            hidden: true,
            results: 6,
            exactTagCount: 2,
            children: [
              {
                title: 'Abnormality of skin morphology (HP:0011121)',
                text: 'Abnormality of skin morphology (HP:0011121)',
                key:
                  'Abnormality of the integument (HP:0001574)-Abnormality of the skin (HP:0000951)-' +
                  'Abnormality of skin morphology (HP:0011121)',
                hidden: true,
                results: 4,
                exactTagCount: 4,
                children: [],
              },
            ],
          },
          {
            title: (
              <span>
                {'Abnormality of skin '}
                <div className="highlight" style={{ display: 'inherit' }}>
                  adnexa
                </div>
                {' morphology (HP:0011138)'}
              </span>
            ),
            text: 'Abnormality of skin adnexa morphology (HP:0011138)',
            key:
              'Abnormality of the integument (HP:0001574)-Abnormality of skin adnexa morphology (HP:0011138)',
            hidden: false,
            results: 6,
            exactTagCount: 1,
            children: [
              {
                title: 'Skin appendage neoplasm (HP:0012842)',
                text: 'Skin appendage neoplasm (HP:0012842)',
                key:
                  'Abnormality of the integument (HP:0001574)-' +
                  'Abnormality of skin adnexa morphology (HP:0011138)-' +
                  'Skin appendage neoplasm (HP:0012842)',
                hidden: true,
                results: 2,
                exactTagCount: 2,
                children: [],
              },
              {
                title: 'Abnormal hair morphology (HP:0001595)',
                text: 'Abnormal hair morphology (HP:0001595)',
                key:
                  'Abnormality of the integument (HP:0001574)-' +
                  'Abnormality of skin adnexa morphology (HP:0011138)-' +
                  'Abnormal hair morphology (HP:0001595)',
                hidden: true,
                results: 2,
                exactTagCount: 2,
                children: [],
              },
              {
                title: 'Custom term xyz',
                text: 'Custom term xyz',
                key:
                  'Abnormality of the integument (HP:0001574)-' +
                  'Abnormality of skin adnexa morphology (HP:0011138)-Custom term xyz',
                hidden: true,
                results: 1,
                exactTagCount: 1,
                children: [],
              },
            ],
          },
        ],
      },
    ];

    const event = {
      target: { value: 'adne' },
    } as React.ChangeEvent<HTMLInputElement>;

    wrapper.find('Input').simulate('change', event);
    expect(wrapper.state().treeData).toEqual(expectedResult);
  });

  it('should only call search in tree for 3 or more characters', () => {
    const event = {
      target: { value: 'ad' },
    } as React.ChangeEvent<HTMLInputElement>;

    wrapper.find('Input').simulate('change', event);
    expect(wrapper.state().expandedKeys).toEqual(['Abnormality of the integument (HP:0001574)']);
  });

  it('should render participants count and exact count for each term', () => {
    const component = mount(
      <SelectionTree
        dataSource={treeData}
        checkedKeys={checkedKeys}
        onItemSelect={onItemSelect}
        targetKeys={targetKeys}
        onItemSelectAll={onItemSelectAll}
        selectedField={''}
      />,
    );

    expect(
      component
        .find('Tag')
        .getElements()
        .map((e) => e.props.children),
    ).toMatchSnapshot();
    component.unmount();
  });

  it('should searchInTree method should return results as expected', () => {
    //Query with from beginning of a term: 'Skin appendage neoplasm (HP:0012842)'
    const resStart = wrapperInstance.instance().searchInTree('Ski', dataSource[0]);
    expect(resStart).toBe(true);
    // Query with from beginning of a term: 'Custom term xyz'
    const resEnd = wrapperInstance.instance().searchInTree('xyz', dataSource[0]);
    expect(resEnd).toBe(true);
    //Query with parenthesis - will be cleaned out'
    const resParenthesis = wrapperInstance.instance().searchInTree('(HP:0012842)', dataSource[0]);
    expect(resParenthesis).toBe(true);
    //Query with terms not existing will return false
    const resNotFound = wrapperInstance.instance().searchInTree('NOTTHERE', dataSource[0]);
    expect(resNotFound).toBe(false);
    //Query for multiple words: 'Abnormality of skin morphology (HP:0011121)'
    const resMultiWords = wrapperInstance.instance().searchInTree('skin mor', dataSource[0]);
    expect(resMultiWords).toBe(true);
  });

  it('should get checked keys for selected items', () => {
    const resStart = wrapperInstance
      .instance()
      .checkKeys(
        'Abnormality of the integument (HP:0001574)-' +
          'Abnormality of skin adnexa morphology (HP:0011138)-' +
          'Skin appendage neoplasm (HP:0012842)',
        treeData,
      );
    expect(resStart).toStrictEqual({
      check: [
        'Abnormality of the integument (HP:0001574)-' +
          'Abnormality of skin adnexa morphology (HP:0011138)-' +
          'Skin appendage neoplasm (HP:0012842)',
      ],
      halfcheck: [
        'Abnormality of the integument (HP:0001574)-' +
          'Abnormality of skin adnexa morphology (HP:0011138)',
        'Abnormality of the integument (HP:0001574)',
      ],
    });
  });
});
