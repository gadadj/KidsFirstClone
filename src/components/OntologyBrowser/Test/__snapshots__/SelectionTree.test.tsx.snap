// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`In SelectionTree should render participants count and exact count for each term 1`] = `
Array [
  0,
  12,
  2,
  6,
  1,
  6,
]
`;
